doc-bot/new-tab-page-messaging.md

```diff
@@ -0,0 +1,376 @@
+            name: "newTabPageMessage"
+        )
+    }
+}
+
+// ✅ CORRECT - Message handling implementation
+extension NewTabPageUserScript: WKScriptMessageHandler {
+    func userContentController(
+        _ userContentController: WKUserContentController,
+        didReceive message: WKScriptMessage
+    ) {
+        guard let body = message.body as? [String: Any],
+              let method = body["method"] as? String,
+              let id = body["id"] as? Int else {
+            return
+        }
+        
+        Task {
+            do {
+                let result = try await handleMethod(method, params: body["params"])
+                sendResponse(result: result, id: id, to: message.webView)
+            } catch {
+                sendError(error, id: id, to: message.webView)
+            }
+        }
+    }
+}
+```
+
+## Content-Scope-Scripts Integration
+
+### Script Injection
+```swift
+// ✅ CORRECT - Content scope script injection
+final class NewTabPageUserContentController {
+    func injectContentScopeScripts() {
+        // Load bundled content-scope-scripts
+        guard let scriptURL = Bundle.main.url(
+            forResource: "newTabPage",
+            withExtension: "js",
+            subdirectory: "content-scope-scripts"
+        ) else { return }
+        
+        let script = try String(contentsOf: scriptURL)
+        let userScript = WKUserScript(
+            source: script,
+            injectionTime: .atDocumentStart,
+            forMainFrameOnly: true
+        )
+        
+        addUserScript(userScript)
+    }
+}
+```
+
+### Frontend API Exposure
+```javascript
+// ✅ CORRECT - Frontend API pattern (from content-scope-scripts)
+window.newTabPage = {
+    widgets: {
+        favorites: {
+            getData: () => sendMessage('favorites_getData'),
+            add: (url) => sendMessage('favorites_add', { url }),
+            delete: (id) => sendMessage('favorites_delete', { id }),
+            reorder: (ids) => sendMessage('favorites_reorder', { ids })
+        },
+        privacyStats: {
+            getData: () => sendMessage('privacyStats_getData'),
+            showDetails: () => sendMessage('privacyStats_showDetails')
+        }
+        // ... other widgets
+    }
+};
+
+function sendMessage(method, params = {}) {
+    return new Promise((resolve, reject) => {
+        const id = generateId();
+        pendingRequests[id] = { resolve, reject };
+        
+        webkit.messageHandlers.newTabPageMessage.postMessage({
+            jsonrpc: '2.0',
+            method,
+            params,
+            id
+        });
+    });
+}
+```
+
+## Message Routing
+
+### Client Registration
+```swift
+// ✅ CORRECT - Dynamic client registration
+final class NewTabPageUserScript {
+    private var clients: [String: any NewTabPageClient] = [:]
+    
+    func registerClient<T: NewTabPageClient>(
+        _ client: T,
+        for prefix: String
+    ) {
+        clients[prefix] = client
+    }
```

Comment:
@afterxleep thanks for preparing this 🙌 I only briefly skimmed over NTP files and I found that the instructions in this file are highly incorrect. We should fix them before merging. I’m not sure if I could look into it tomorrow, but worst case we could exclude NTP files from this PR before merging.

Reply:
@ayoy I generated this using AI as a starting point - May have used a 'cheap' model though.. Feel free to review and update the Md's manually or remove them it don't make sense at all.

Reply:
@ayoy I've removed the NTP files.  so we can get this moving forward.  We can always add later

Reply:
@afterxleep thanks, that sounds best, sorry I couldn’t get to it in time.